src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp

```diff
@@ -0,0 +1,290 @@
+#ifdef ENABLE_ANNOY
+
+#include <Storages/MergeTree/MergeTreeIndexAnnoy.h>
+
+#include <Common/typeid_cast.h>
+#include <Core/Field.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+
+
+namespace DB
+{
+
+namespace Annoy
+{
+
+template<typename Dist>
+void AnnoyIndexSerialize<Dist>::serialize(WriteBuffer& ostr) const
+{
+    if (!Base::_built)
+    {
+        throw Exception("Annoy Index should be built before serialization", ErrorCodes::LOGICAL_ERROR);
+    }
+    writeIntBinary(Base::_s, ostr);
+    writeIntBinary(Base::_n_items, ostr);
+    writeIntBinary(Base::_n_nodes, ostr);
+    writeIntBinary(Base::_nodes_size, ostr);
+    writeIntBinary(Base::_K, ostr);
+    writeIntBinary(Base::_seed, ostr);
+    writeVectorBinary(Base::_roots, ostr);
+    ostr.write(reinterpret_cast<const char*>(Base::_nodes), Base::_s * Base::_n_nodes);
+}
+
+template<typename Dist>
+void AnnoyIndexSerialize<Dist>::deserialize(ReadBuffer& istr)
+{
+    readIntBinary(Base::_s, istr);
+    readIntBinary(Base::_n_items, istr);
+    readIntBinary(Base::_n_nodes, istr);
+    readIntBinary(Base::_nodes_size, istr);
+    readIntBinary(Base::_K, istr);
+    readIntBinary(Base::_seed, istr);
+    readVectorBinary(Base::_roots, istr);
+    Base::_nodes = realloc(Base::_nodes, Base::_s * Base::_n_nodes);
+    istr.read(reinterpret_cast<char*>(Base::_nodes), Base::_s * Base::_n_nodes);
+
+    Base::_fd = 0;
+    // set flags
+    Base::_loaded = false;
+    Base::_verbose = false;
+    Base::_on_disk = false;
+    Base::_built = true;
+}
+
+template<typename Dist>
+float AnnoyIndexSerialize<Dist>::getSpaceDim() const
+{
+    return Base::get_f();
+}
+
+}
+
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int INCORRECT_QUERY;
+}
+
+MergeTreeIndexGranuleAnnoy::MergeTreeIndexGranuleAnnoy(const String & index_name_, const Block & index_sample_block_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+    , index_base(nullptr)
+{}
+
+MergeTreeIndexGranuleAnnoy::MergeTreeIndexGranuleAnnoy(
+    const String & index_name_,
+    const Block & index_sample_block_,
+    AnnoyIndexPtr index_base_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+    , index_base(std::move(index_base_))
+{}
+
+bool MergeTreeIndexGranuleAnnoy::empty() const
+{
+    return !static_cast<bool>(index_base);
+}
+
+void MergeTreeIndexGranuleAnnoy::serializeBinary(WriteBuffer & ostr) const
+{
+    writeIntBinary(index_base->getSpaceDim(), ostr); // write dimension
+    index_base->serialize(ostr);
+}
+
+void MergeTreeIndexGranuleAnnoy::deserializeBinary(ReadBuffer & istr, MergeTreeIndexVersion /*version*/)
+{
+    int dimension;
+    readIntBinary(dimension, istr);
+    index_base = std::make_shared<AnnoyIndex>(dimension);
+    index_base->deserialize(istr);
+}
+
+
+MergeTreeIndexAggregatorAnnoy::MergeTreeIndexAggregatorAnnoy(const String & index_name_,
```

Comment:
Cosmetic: Please either put everything on one line or use tab indentation like that:

```
MergeTreeIndexAggregatorAnnoy::MergeTreeIndexAggregatorAnnoy(
    const String & index_name_,
    const Block & index_sample_block_,
    int index_param_)
{
```